nixos/modules/services/misc/anki-sync-server.md

```diff
@@ -0,0 +1,68 @@
+# Anki Sync Server {#module-services-anki-sync-server}
+
+[Anki Sync Server](https://docs.ankiweb.net/sync-server.html) is the built-in
+sync server, present in recent versions of Anki. Advanced users who cannot or
+do not wish to use AnkiWeb can use this sync server instead of AnkiWeb.
+
+This module is compatible only with Anki versions >=2.1.66, due to [recent
+enhancements to the Nix anki
+package](https://github.com/NixOS/nixpkgs/commit/05727304f8815825565c944d012f20a9a096838a).
+
+## Basic Usage {#module-services-anki-sync-server-basic-usage}
+
+By default, the module creates a
+[`systemd`](https://www.freedesktop.org/wiki/Software/systemd/)
+unit which runs the sync server with an isolated user using the systemd
+`DynamicUser` option.
+
+This can be done by enabling the `anki-sync-server` service:
+```
+{ ... }:
+
+{
+  services.anki-sync-server.enable = true;
+}
+```
+
+It is necessary to set at least one username-password pair under
+{option}`services.anki-sync-server.users`. For example
+
+```
+{
+  services.anki-sync-server.users = [
+    {
+      username = "user";
+      passwordFile = /etc/anki-sync-server/user;
+    }
+  ];
+}
+```
+
+Here, `passwordFile` is the path to a file containing just the password in
+plaintext. Make sure to set permissions to make this file unreadable to any
+user besides root.
+
+By default, the server listen address {option}`services.anki-sync-server.host`
+is set to localhost, listening on port
+{option}`services.anki-sync-server.port`, and does not open the firewall. This
+is suitable for purely local testing, or to be used behind a reverse proxy. If
+you want to expose the sync server directly to other computers (not recommended
+in most circumstances, because the sync server doesn't use HTTPS), then set the
+following options:
+
+```
+{
+  services.anki-sync-server.host = "0.0.0.0";
+  services.anki-sync-server.openFirewall = true;
+}
+```
+
+
+## Alternatives {#module-services-anki-sync-server-alternatives}
+
+The [`ankisyncd` NixOS
+module](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/ankisyncd.nix)
+provides similar functionality, but using a third-party implementation,
+[`anki-sync-server-rs`](https://github.com/ankicommunity/anki-sync-server-rs/).
+According to that project's README, it is "no longer maintained", and not
+recommended for Anki 2.1.64+.
```

Comment:
I'm curious where did you see that? It works fine with 2.1.66 and I already have it patched for the 23.10 beta that does require an update; so it seems fine to use as far as I'm concerned:
https://github.com/ankicommunity/anki-sync-server-rs/pull/76
(I haven't submitted it to nixpkgs yet because it requires https://github.com/ankitects/rust-csv/pull/1 or equivalent to build anki itself and I didn't get any feedback from upstream yet; but I have it all ready along with a trivial nixosTest...)

We were discussing the future of anki-sync-server-rs vs the built-in server in an old PR here: https://github.com/NixOS/nixpkgs/pull/224366#issuecomment-1714697175

I guess since you've done the work now we can keep both in parallel for a while; I was mostly curious about the compatibility of the server data since users aren't too bad to configure again.

Reply:
Ah, I got that impression based on the anki-sync-server-rs README: https://github.com/ankicommunity/anki-sync-server-rs/blob/5280f9f2034b29d77388df497ebcdb7d13cb5c94/README.md

Reply:
For now, we should probably keep them both, until we figure out how to migrate data between the two servers in a less painful way.

Reply:
Right, that note in the README is what got us started in the nixpkgs thread -- I didn't realize it also explicitly mentioned 1.64.

As far as the protocol (and thus clients supported) go, anki-sync-server-rs uses the code from anki itself so they support exactly the same thing.
As long as someone keeps rebasing the small anki patch in anki-sync-server-rs I don't expect any breakage in the short-medium term; I've spent much more time getting anki 23.10beta1 to build on nixpkgs (because of the cargo vendor thing) than updating the standalone server patch :|

Long-term though I agree the standalone server is fragile so migrating to this makes more sense; when writing the nixosTest for ankisyncd I also painfully realized how non-nixos-y the old server is (having to run a command manually in the data dir to manage users), but I'd really like to be able to preserve data itself.

I'll try to run the built in anki server and compare data directories this weekend

Reply:
So the build-in server data seems compatible: just need to set the base path (SYNC_BASE) to /var/lib/ankisyncd/collections and it rolls.
The login token isn't compatible -- I get an auth error on switching from ankisyncd to anki's server and back, but data itself seems fine, which is what's most important.

With that I guess we really ought to make the switch, there's really no point in keeping both... I have mixed feelings as I just did the port to anki-sync-server-rs but rationally it's definitely for the best. There have been no release with the new server, so I'd go ahead and replace it, just keeping the SYNC_BASE for convenience.
(sorry for the delay, and rushing this in after it...)

I'll see about packaging the standalone server either this week or next week; anki 23.10 should now build fine with https://github.com/ankitects/anki/pull/2703 hopefully.
I'll also test this PR a bit more thoroughly, I've never used systemd's LoadCredential so I just don't get it without testing, but if it works it looks good to me.

Reply:
Ok on LoadCredential; this is more complicated than I'd do (I'd just ask users to fill in an EnvironmentFile themselves), but it works and is certainly more nixos-y so that's fine with me